pkg/dyninst/decode/decoder.go

```diff
@@ -0,0 +1,407 @@
+			decoder.typesToProbesAndEvents[event.Type.ID] = probeEvent{
+				event: event,
+				probe: probe,
+			}
+		}
+	}
+	return decoder, nil
+}
+
+type dataItemWithAddrCounter struct {
+	dataItem    output.DataItem
+	addrCounter uint8
+}
+
+// Decode decodes the given event into the given writer.
+func (d *Decoder) Decode(event output.Event, out io.Writer) error {
+	var (
+		rootData  []byte
+		eventRoot *ir.EventRootType
+		ok        bool
+		err       error
+		frames    []uint64
+	)
+
+	frames, err = event.StackPCs()
+	if err != nil {
+		return err
+	}
+	enc := jsontext.NewEncoder(out)
+	err = enc.WriteToken(jsontext.BeginObject)
+	if err != nil {
+		return err
+	}
+
+	err = enc.WriteToken(jsontext.String("stack_frames"))
+	if err != nil {
+		return err
+	}
+	err = enc.WriteToken(jsontext.BeginArray)
+	if err != nil {
+		return err
+	}
+	for i := range frames {
+		err = enc.WriteToken(jsontext.Uint(frames[i]))
+		if err != nil {
+			return err
+		}
+	}
+	err = enc.WriteToken(jsontext.EndArray)
+	if err != nil {
+		return err
+	}
+
+	err = enc.WriteToken(jsontext.String("captures"))
+	if err != nil {
+		return err
+	}
+	err = enc.WriteToken(jsontext.BeginObject)
+	if err != nil {
+		return err
+	}
+	itemsByAddress := map[uint64]*dataItemWithAddrCounter{}
+	for item, err := range event.DataItems() {
+		if errors.Is(err, output.ErrFinishedIterating) {
+			break
+		}
+		if err != nil {
+			return err
+		}
+		if eventRoot == nil {
+			// item is EventRoot
+			rootData = item.Data()
+			eventRoot, ok = d.program.Types[ir.TypeID(item.Header().Type)].(*ir.EventRootType)
+			if !ok {
+				return errors.New("expected event of type root first")
+			}
+			continue
+		}
+
+		itemsByAddress[item.Header().Address] = &dataItemWithAddrCounter{
+			dataItem:    item,
+			addrCounter: 0,
+		}
+	}
+
+	// We iterate over the 'Expressions' of the EventRoot which contains
+	// metadata and raw bytes of the parameters of this function.
+	for _, expr := range eventRoot.Expressions {
+		parameterType := expr.Expression.Type
+		parameterData := rootData[expr.Offset : expr.Offset+parameterType.GetByteSize()]
+		err = enc.WriteToken(jsontext.String(expr.Name))
+		if err != nil {
+			return err
+		}
+		err = d.encodeValue(enc, itemsByAddress, parameterType, parameterData)
+		if err != nil {
+			return fmt.Errorf("error parsing data for field %s: %w", eventRoot.Name, err)
+		}
+	}
+
+	err = enc.WriteToken(jsontext.EndObject) // captures
+	if err != nil {
+		return err
+	}
+	err = enc.WriteToken(jsontext.EndObject) // full object
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (d *Decoder) encodeValue(enc *jsontext.Encoder, itemsByAddress map[uint64]*dataItemWithAddrCounter, irType ir.Type, data []byte) error {
+	switch v := irType.(type) {
+	case *ir.BaseType:
+		return d.encodeBaseTypeValue(enc, v, data)
+	case *ir.PointerType:
+		if len(data) < int(v.GetByteSize()) {
+			return errors.New("passed data not long enough for pointer")
+		}
+		// handle nil pointer
```

Comment:
nit: does this comment add a lot of value? Also it feels slightly misplaced. I could see it next to the `if` as an inline comment.